### PR TITLE
Session 0 Ruby Questions

### PR DESCRIPTION
**Pull request by ronithrnair**
Session 0 Ruby questions completed (I think)

### CODE REVIEW

**Comment by abhishekkumar2718** [comment]
As you can see by the failing check for Session 0, ruby is _not_ done - try running the tests locally to check if your solutions are correct

**Comment by ronithrnair** [comment]
Sorry, I forgot to merge a branch into the session_0 branch

**Comment by ronithrnair** [comment]
> As you can see by the failing check for Session 0, ruby is _not_ done - try running the tests locally to check if your solutions are correct

The tests ran properly now that I've corrected my code.

File: session_0/fizz_buzz.rb [hunk 1]
```diff
@@ -36,6 +36,27 @@
...
 # their function arguments were incorrect.
-def fizz_buzz(n:, x: 3, y: 5)
-  raise NotImplementedError # TODO
-end
+def fizz_buzz(n:, x: 3, y: 5 )
+  if (n < 0) || (x<= 0) || (y<=0) then
+     raise ArgumentError # TODO
+    end
+   a = []
+   for i in 1..n do
```
Comment (abhishekkumar2718) [inline session_0/fizz_buzz.rb:44]: Do you think map function can be of use here?

https://www.rubyguides.com/2018/10/ruby-map-method/

File: session_0/item.rb [hunk 2]
```diff
@@ -23,14 +23,28 @@ def initialize(name: '', price: 0, quantity: 1, category_id: 4,
   # Returns a boolean value whether than item is discounted i.e. the
   # discount deadline has been crossed or not.
   def discounted?
-    raise NotImplementedError # TODO
+    if (Time.now > @discount_deadline)
```
Comment (abhishekkumar2718) [inline session_0/item.rb:26]: - Use `discount_deadline` as we have define attr accessor for it.

```ruby
if condition
  return false
else
  return true
end
```

is same as:
```ruby
return !condition
```

File: session_0/item.rb [hunk 3]
```diff
@@ -23,14 +23,28 @@ def initialize(name: '', price: 0, quantity: 1, category_id: 4,
...
+    end
   end
 
   # If the item is discounted, the current price is 
   # `price * (100 - discount rate) / 100`. Otherwise, it is same as the price.
   #
   # TODO: Implement instance method 'current_price'
 
+  def current_price
+    if discounted? then
```
Comment (abhishekkumar2718) [inline session_0/item.rb:40]: Don't use the `then` keyword, it does not add to the readability.

File: session_0/item.rb [hunk 4]
```diff
@@ -41,5 +55,16 @@ def discounted?
   # Note: If there are no items for category, stock price for category
   # should be zero.
   def self.stock_price_by_category(items)
+    total_category_price = {
```
Comment (abhishekkumar2718) [inline session_0/item.rb:58]: What if there are 100 categories? Would writing a hash by hand still make sense?

File: session_0/word_processing.rb [hunk 5]
```diff
@@ -2,13 +2,16 @@
...
 def lower_case(words)
-  raise NotImplementedError # TODO
+  return words.map { |a| a.downcase } # TODO
 end
 
 # Similar to `lower_case`, this function modifies the array in-place
 # and does not return any value.
 def lower_case!(words)
-  raise NotImplementedError # TODO
+  for a in words
```
Comment (abhishekkumar2718) [inline session_0/word_processing.rb:11]: Use `words.each` instead of a `for` loop

File: session_0/word_processing.rb [hunk 6]
```diff
@@ -34,11 +37,32 @@ def words_with_prefix(prefix, words)
 # The function `similarity_score` takes two words and returns the
 # similarity score (an integer).
 def similarity_score(word_1, word_2)
-  raise NotImplementedError # TODO
+  size = [word_1.size,word_2.size].min # TODO
+  score = 0
+  for i in 0..size-1
```
Comment (abhishekkumar2718) [inline session_0/word_processing.rb:42]: What are you iterating over in the for loop?

**Review by abhishekkumar2718 — CHANGES_REQUESTED** [review]

File: session_0/word_processing.rb (hunk 6 above)
Comment (ronithrnair) [reply session_0/word_processing.rb:42]: I was trying to extract each individual character in the word in the same way one does in C++. Is it possible to use word_1.each  to get individual characters?

**Comment by ronithrnair** [comment]
I changed Fizz Buzz to use map and I removed all the 'then's. I committed these changes - What do I do now ?  As in , can you see the changes that I have made?  Also I don't know how to resolve the other issue -  the writing by hash one.

File: session_0/word_processing.rb (hunk 6 above)
Comment (abhishekkumar2718) [reply session_0/word_processing.rb:40]: Experiment and find out

File: session_0/fizz_buzz.rb [hunk 7]
```diff
@@ -36,6 +36,23 @@
...
+     str << "Fizz"
+     end
+     if i % y == 0 
+     str << "Buzz"
+     end
+     
+     if str== "" 
+     str << i.to_s
+     end
+     k = str
```
[the remark this answers is not in the document]
Comment (ronithrnair) [reply session_0/fizz_buzz.rb:56]: It was a mistake, really. I forgot to remove the statement from a previous version of the code. I now realise that the presence of an assignment statement somehow made the program work. From what I understand it has to do with the fact that assignment statements return the value of whatever value it's assigning.

**Comment by ronithrnair** [comment]
I think I made all the changes.

File: session_0/word_processing.rb [hunk 8]
```diff
@@ -34,11 +35,45 @@ def words_with_prefix(prefix, words)
...
```
Comment (abhishekkumar2718) [inline session_0/word_processing.rb:40]: See if you can use `find` function here: https://apidock.com/ruby/Enumerable/find

**Review by abhishekkumar2718 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**